C codegen bug: name mangling collision with nested tuples of matching flattened types

_See https://github.com/nim-lang/Nim/issues/25883#issuecomment-4658536587 for a much smaller repro without macros. Tuples name mangling doesn't handle static[T] properly_

***

The following generates wrong C code:

```Nim
import std/macros

type
  Int[V: static int] = object
  Layout[Sh, St] = object
    shape*: Sh
    stride*: St

template make_layout(shapeArg, strideArg: untyped): untyped =
  Layout[typeof(shapeArg), typeof(strideArg)](
    shape: shapeArg, stride: strideArg
  )

macro group(layout: Layout; B, E: static int): untyped =
  let lTyp = layout.getTypeInst()
  let shTyp = lTyp[1]
  let R =
    if shTyp.kind == nnkTupleConstr:
      shTyp.len
    else:
      1
  var ct: tuple[shape, stride: seq[NimNode]]
  for i in 0 ..< B:
    ct.shape.add nnkBracketExpr.newTree(nnkDotExpr.newTree(layout, ident"shape"), newLit i)
    ct.stride.add nnkBracketExpr.newTree(nnkDotExpr.newTree(layout, ident"stride"), newLit i)
  var gSh = newNimNode(nnkTupleConstr)
  var gSt = newNimNode(nnkTupleConstr)
  for i in B ..< E:
    gSh.add nnkBracketExpr.newTree(nnkDotExpr.newTree(layout, ident"shape"), newLit i)
    gSt.add nnkBracketExpr.newTree(nnkDotExpr.newTree(layout, ident"stride"), newLit i)
  ct.shape.add gSh; ct.stride.add gSt
  for i in E ..< R:
    ct.shape.add nnkBracketExpr.newTree(nnkDotExpr.newTree(layout, ident"shape"), newLit i)
    ct.stride.add nnkBracketExpr.newTree(nnkDotExpr.newTree(layout, ident"stride"), newLit i)
  var outSh = newNimNode(nnkTupleConstr)
  var outSt = newNimNode(nnkTupleConstr)
  for i in 0 ..< ct.shape.len:
    outSh.add ct.shape[i]
    outSt.add ct.stride[i]
  result = newCall(bindSym"make_layout", outSh, outSt)

  echo result.toStrLit()

proc main() =
  let a = make_layout((Int[2](), Int[3](), Int[5](), Int[7]()),
                      (Int[1](), Int[2](), Int[6](), Int[30]()))

  let b = a.group(0, 2)
  let c = a.group(0, 3)

main()
```

```
/@mnimbug_group_c_struct_truncation.nim.c: In function ‘main__nimbug95group95c95struct95truncation_u94’:
/@mnimbug_group_c_struct_truncation.nim.c:147:26: error: ‘tyTuple__SY4Zk0uAPIqtr88aujunWw’ has no member named ‘Field2’; did you mean ‘Field0’?
  147 |         c_1.shape.Field0.Field2 = a_1.shape.Field2;
      |                          ^~~~~~
      |                          Field0
/@mnimbug_group_c_struct_truncation.nim.c:148:28: error: incompatible types when assigning to type ‘tyObject_Int__PU6XOYxqzl13wDO1MIp8HA’ from type ‘tyObject_Int__Hb7AvaBFX1sV0DFpl9aH7MA’
  148 |         c_1.shape.Field1 = a_1.shape.Field3;
      |                            ^~~
/@mnimbug_group_c_struct_truncation.nim.c:151:27: error: ‘tyTuple__kea9bgY0r9a65GAy4HPbWI1Q’ has no member named ‘Field2’; did you mean ‘Field0’?
  151 |         c_1.stride.Field0.Field2 = a_1.stride.Field2;
      |                           ^~~~~~
      |                           Field0
```

--- a/compiler/sighashes.nim
+++ b/compiler/sighashes.nim
@@ -211,6 +211,7 @@ proc hashType(c: var MD5Context, t: PType; flags: set[ConsiderFlag]; conf: Confi
     c.hashTree(t.n, {}, conf)
   of tyTuple:
     c &= char(t.kind)
+    c &= t.len
     if t.n != nil and CoType notin flags:
       for i in 0..<t.n.len:
         assert(t.n[i].kind == nkSym)

--- a/tests/ccgbugs2/t25883_tuple_sighash_collision.nim
+++ b/tests/ccgbugs2/t25883_tuple_sighash_collision.nim
@@ -1,4 +1,5 @@
 discard """
+targets: "c cpp"
 output: "13"
 """
 
@@ -16,3 +17,29 @@ proc main() =
   echo c[0][2]
 
 main()
+
+block:
+  type
+    Int[V: static int] = object
+    Layout[Sh, St] = object
+      shape: Sh
+      stride: St
+  
+  func makeB(): auto =
+    Layout[((Int[2], Int[3]), Int[5], Int[7]), ((Int[1], Int[2]), Int[6], Int[30])](
+      shape: ((Int[2](), Int[3]()), Int[5](), Int[7]()),
+      stride: ((Int[1](), Int[2]()), Int[6](), Int[30]())
+    )
+  
+  func makeC(): auto =
+    Layout[((Int[2], Int[3], Int[5]), Int[7]), ((Int[1], Int[2], Int[6]), Int[30])](
+      shape: ((Int[2](), Int[3](), Int[5]()), Int[7]()),
+      stride: ((Int[1](), Int[2](), Int[6]()), Int[30]())
+    )
+  
+  
+  proc main() =
+    let b = makeB()
+    let c = makeC()
+  
+  main()

--- a/tests/ccgbugs2/t25883_tuple_sighash_collision.nim
+++ b/tests/ccgbugs2/t25883_tuple_sighash_collision.nim
@@ -1,5 +1,5 @@
 discard """
-output: "5"
+output: "13"
 """
 
 # bug #25883: C codegen assigns same type hash to tuples with different nesting

--- a/tests/ccgbugs2/t25883_tuple_sighash_collision.nim
+++ b/tests/ccgbugs2/t25883_tuple_sighash_collision.nim
@@ -1,0 +1,18 @@
+discard """
+output: "5"
+"""
+
+# bug #25883: C codegen assigns same type hash to tuples with different nesting
+# but identical flattened content.
+# ((Int[1], Int[2]), Int[13], Int[14]) and ((Int[1], Int[2], Int[13]), Int[14])
+# must get distinct C type names.
+
+type
+  Int[V: static int] = object
+
+proc main() =
+  var b = ((1, 2), 13, 14)
+  var c = ((1, 2, 13), 14)
+  echo c[0][2]
+
+main()

